Fix markdown editor reference link having colon issue

mushtaqak opened the pull request:
[TNL-2609](https://openedx.atlassian.net/browse/TNL-2609)

This PR fixes the markdown editor issue of reference link having a colon in it. Actually http://example.com/colon:test was converted to http://example.com/colon%3atest

mushtaqak commented:
test this please

mushtaqak commented:
jenkins run python

mushtaqak commented:
jenkins run python

mushtaqak commented:
@jimabramson @cahrens @waheedahmed Please review.

cahrens commented on common/test/acceptance/tests/discussion/test_discussion.py:
```diff
@@ -236,6 +236,22 @@ def test_mathjax_rendering(self):
         self.assertTrue(self.thread_page.is_mathjax_rendered())
 
+    def test_markdown_reference_link(self):
+        sample_link = "http://example.com/colon:test"
```
I prefer that tests have documentation explaining what they are testing. Our best practice is to use BDD-style test descriptions, but at the least, add a documentation block.

cahrens commented on lms/static/js/Markdown.Converter.js:
```diff
@@ -1285,27 +1285,12 @@ else
         //  attacklab: Utility functions
         //
 
-        var _problemUrlChars = /(?:["'*()[\]:]|~D)/g;
```
Why did you decide to delete this method (are we not concerned with some of these "problem" chars)?

mushtaqak replied:
@cahrens Because this method has been removed in the latest version of the Markdown Library. The changes I have done are from the latest version Markdown library; picking the snippet that was used to solve this issue.

mushtaqak replied:
Actually, we could have alternatively updated the library which contains three files but one of them has changes at many places.

cahrens replied:
I'd prefer us to upgrade the Markdown library instead of bringing in bits of a newer version.

Looking at the commit history of Markdown.Converter.js, we haven't made any modifications to this file since 7/25/2012 (we just moved it a couple of places). To be entirely confident that we haven't modified this file from its original form, it would be great to find the original someplace and diff the file.

cahrens replied:
Looks like our comments crossed... as I noted below, I'm OK with the targeted change you are making, but would like comments in the code explaining where you got the newer method from.

jimabramson commented on lms/static/js/Markdown.Converter.js:
```diff
@@ -1329,4 +1314,4 @@ else
     }; // end of the Markdown.Converter constructor
 
-})();
+})();
```
minor: revert this last line

jimabramson commented on lms/static/js/Markdown.Converter.js:
```diff
@@ -1285,27 +1285,12 @@ else
-            });
+        function attributeSafeUrl(url) {
+            url = attributeEncode(url);
+            url = escapeCharacters(url, "*_:()[]")
```
question: why are `()[]` included here?

cahrens commented:
I don't have much context on this change... honestly wasn't aware that we use this markdown editor. Glad to see that @jimabramson is looking at it!

cahrens commented:
Actually, it looks like we have edited Markdown.Editor.js, so an upgrade probably is out of scope for this bug fix. We should have a technical debt story though to update this vendor code.

@mushtaqak I am OK with the targeted fix, but please add a comment in the code explaining where you got the new code from.

jimabramson commented:
@mushtaqak Do you mind explaining what causes the links not to work as they are?

`:` isn't even technically a legal character in the path part of a URL, but our system tolerates it.  However, I've tested with urlencoded course keys in courseware urls (e.g. "course-v1:EPFLx+TrigoExpX+2T2015" -> "course-v1%3AEPFLx%2BTrigoExpX%2B2T2015") and either form works for me.  If anything I would think that we would want to fix the URLs that cannot accept urlencoded path parts, instead of skipping the urlencoding.

mushtaqak commented:
@cahrens @jimabramson  I have addressed your suggestions and added where I got the code fixing snippet.

@jimabramson I think `:` is allowed in the URL path, [see W3 specs](http://www.w3.org/TR/html4/types.html#type-name), however Markdown had a bug in their previous versions that escaped `:` to `%3a` which they have fixed in recent versions.

cahrens commented:
:+1:

jimabramson commented:
@mushtaqak I was thinking of this bit from [rfc 2396](http://www.ietf.org/rfc/rfc2396.txt):

> 2.2. Reserved Characters
> 
>   Many URI include components consisting of or delimited by, certain
>   special characters.  These characters are called "reserved", since
>   their usage within the URI component is limited to their reserved
>   purpose.  If the data for a URI component would conflict with the
>   reserved purpose, then the conflicting data must be escaped before
>   forming the URI.
>
>      reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" |
>                    "$" | ","
>
>   The "reserved" syntax class above refers to those characters that are
>   allowed within a URI, but which may not be allowed within a
>   particular component of the generic URI syntax; they are used as
>   delimiters of the components described in Section 3.

I still don't understand why the courseware can't handle a link with a urlencoded course id, maybe I am still missing something?  If there are parts of courseware that can't handle such requests it seems like we should consider addressing those.

Meanwhile, since we are backporting a fix from the offical Markdown library in order to solve a specific bug, I don't want to block that process.  :+1:

mushtaqak commented:
jenkins run python

mushtaqak commented:
@jimabramson while this PR addresses markdown link issue which solves the issue; however you are right, I just checked forum discussion module and other courseware ,  the url with colon in it fails on forum discussion single thread view.

mushtaqak commented:
jenkins test python

mushtaqak commented:
jenkins run python

mushtaqak commented:
@adampalay FYI

mushtaqak commented:
@jimabramson Are we good to merge ?

jimabramson commented:
@mushtaqak yes. :+1:

cahrens commented:
@mushtaqak what is the ticket to fix the server-side issue with the colon? I'm confused about what this PR accomplishes, since you said that the url with the colon fails on the forum discussion single thread view. What is the value of the client side working when the server side still does not?

mushtaqak commented:
@jimabramson @cahrens Actually this PR fixes the issue in the ticket that users are unable to add correct links to the posts with colon in the markdown editor.

While after discussions above, I found another case that if a user enters a discussion thread directly in the browser with colon, it is not encoded correctly.

jimabramson commented:
OK I'm back to my earlier position on this.  if we are backporting a specific fix from a newer version of Markdown, and that solves the user-facing bug, then I'm OK with merging.  We can address the failure to handle urlencoded course ids in a separate ticket.  :+1:  :+1:

jimabramson commented:
@mushtaqak FYI I've done a little investigation into the forums urlencoding problem, though I haven't found the root cause I've confirmed that if the course_id has any urlencoded parts in it, then [this function](https://github.com/edx/edx-platform/blob/master/common/static/coffee/src/discussion/discussion_router.coffee#L44) doesn't get called.  Otherwise it would get called.

mushtaqak commented:
For records; lets add this comment as well, from the HipChat conversation :

Actually Markdown would still need to be fixed as user may also add a link other than edx course url , in that case it may/may not work (e.g. user may add a link : http:// example.com/path:something which will be converted to http:// example.com/path%3asomething) ; So it would depend upon the user entered link and it's url encoding system. I think that we would not want to depend upon others, As Markdown had also fixed that issue, so I think we would still need markdown fix anyway.
